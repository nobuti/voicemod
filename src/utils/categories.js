export default (voices) => {
  const categories = voices.reduce((memo, voice) => {
    return voice.tags != null ? memo.concat(voice.tags) : memo;
  }, []);

  const unique = [...new Set(categories)].map((c) => ({ label: c, value: c }));
  return [{ label: "All", value: "all" }].concat(unique);
};
