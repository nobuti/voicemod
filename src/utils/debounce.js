export default (func, delay) => {
  let inDebounce;
  return function () {
    const args = arguments;
    clearTimeout(inDebounce);
    inDebounce = setTimeout(() => func.apply(this, args), delay);
  };
};
