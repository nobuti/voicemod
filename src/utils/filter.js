const sort = ({ voices, direction }) => {
  if (!["desc", "asc"].includes(direction)) {
    return voices;
  }

  return direction === "asc"
    ? voices.sort((a, b) => a.name.localeCompare(b.name))
    : voices.sort((a, b) => b.name.localeCompare(a.name));
};

const filter = ({ voices, categorie }) => {
  return voices.filter((voice) => voice.tags.includes(categorie));
};

const search = ({ voices, query }) =>
  voices.filter((voice) => voice.name.includes(query));

const applyFilter = ({ voices, options }) => {
  let result = [...voices];

  if (options.search !== "") {
    result = search({ voices: result, query: options.search });
  }

  if (options.categorie !== "all") {
    result = filter({ voices: result, categorie: options.categorie });
  }

  if (options.sort) {
    result = sort({ voices: result, direction: options.sort });
  }

  return result;
};

export { sort, filter, search, applyFilter };
