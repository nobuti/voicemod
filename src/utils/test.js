import React from "react";
import configureStore from "redux-mock-store";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

const mockStore = configureStore();
const mockedStore = (mockedState) => {
  return mockStore(mockedState);
};

export const customRender = (ui, mockedState = null) => {
  const store = mockedStore(mockedState);

  const WithProvider = ({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return { ...render(ui, { wrapper: WithProvider }) };
};
