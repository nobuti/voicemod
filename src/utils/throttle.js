export default (callback, limit) => {
  let timeoutHandler = null;

  return (...args) => {
    if (timeoutHandler == null) {
      timeoutHandler = setTimeout(() => {
        callback(...args);
        timeoutHandler = null;
      }, limit);
    }
  };
};
