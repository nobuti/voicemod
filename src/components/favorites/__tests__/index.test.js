import React from "react";
import { screen } from "@testing-library/react";

import { customRender } from "../../../utils/test";
import Favorites from "../";

describe("Favorites", () => {
  test("should render properly", () => {
    const props = {
      voices: [
        { id: "wadus", name: "Wadus", icon: "VoicesVoiceIcon01.png" },
        { id: "manzana", name: "Manzana", icon: "VoicesVoiceIcon02.png" },
        { id: "cebolleta", name: "Cebolleta", icon: "VoicesVoiceIcon01.png" },
      ],
      favorites: ["manzana"],
    };

    const { container } = customRender(<Favorites {...props} />, {
      voices: {
        collection: props.voices,
        favorites: props.favorites,
      },
    });

    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("grid-title").textContent).toEqual(
      "Favourite voices"
    );
    expect(screen.getAllByTestId("voice")).toHaveLength(props.favorites.length);
  });

  test("should show empty message properly", () => {
    const props = {
      voices: [
        { id: "wadus", name: "Wadus", icon: "VoicesVoiceIcon01.png" },
        { id: "manzana", name: "Manzana", icon: "VoicesVoiceIcon02.png" },
        { id: "cebolleta", name: "Cebolleta", icon: "VoicesVoiceIcon01.png" },
      ],
      favorites: [],
    };

    customRender(<Favorites {...props} />, {
      voices: {
        collection: props.voices,
        favorites: props.favorites,
      },
    });

    expect(screen.getByTestId("favourite-empty")).toBeInTheDocument();
  });
});
