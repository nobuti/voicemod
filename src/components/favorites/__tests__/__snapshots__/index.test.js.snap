// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should render properly 1`] = `
<div>
  <section
    class="Section"
    data-testid="grid"
  >
    <h1
      class="Heading"
      data-testid="grid-title"
    >
      Favourite voices
    </h1>
    <ul
      class="Grid"
      data-testid="grid-list"
    >
      <li
        style="opacity: 0; transform: scale(0.7);"
      >
        <div
          class="Voice is-favorite"
          data-testid="voice"
          tabindex="0"
        >
          <div
            class="Voice-image"
          >
            <img
              alt="Manzana"
              src="/VoicesVoiceIcon02.png"
            />
            ;
          </div>
          <div
            class="Voice-name"
          >
            Manzana
          </div>
          <button
            class="Voice-favorite"
            data-testid="voice-fav"
          >
            Please don't
          </button>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
